### DISCUSSION TITLE
Unable to load fips provider in second load getting init fail error

### DISCUSSION THREAD

**Discussion by absrivas-in** [openssl/openssl#25466] [Q&A]
We are encountering an error while loading the` fips.so` module in a specific use case. The issue arises when using OpenSSL 3.0.15.
We are building a static library (SDK) that contains SDK-related APIs, along with the object files of OpenSSL's libcrypto.a and libssl.a. This allows us to deliver the SDK without including the `libcrypto.a `and `libssl.a` binaries.

The SDK provides a public `load_fips` API that loads the OpenSSL FIPS provider.

**Use Case:**
1. The consumer calls the `load_fips` API from the main process.
2. The consumer then calls the `load_fips` API again, but this time from their own shared library, which is loaded using `dlopen`.

During the second attempt to load fips.so, an error occurs at this line:
[provider_seeding.c#L30](https://github.com/openssl/openssl/blob/openssl-3.0/providers/common/provider_seeding.c#L30)
The error is triggered by a call from:
[fipsprov.c#L551](https://github.com/openssl/openssl/blob/openssl-3.0/providers/fips/fipsprov.c#L551)

**Reason for Failure:** 

During the first FIPS provider load, a global variable [c_get_entropy](https://github.com/openssl/openssl/blob/master/providers/common/provider_seeding.c#L14C34-L14C47) is assigned a function pointer. 
However, when attempting to load the FIPS provider a second time, the function pointer does not match, causing the initialization to fail (`init error`). This behavior is caused by the following sanity check:

`#define set_func(c, f) if (c == NULL) c = f; else if (c != f) return 0;`
As a result, the second attempt to load the FIPS provider fails with an error.

**Sample Logs:**

> Using OpenSSL libs from: openssl/lib/ossl-modules
> Loading base provider...
> Loading fips provider...
> All providers have been loaded!
> call_load_crypto 1
> 
> _init_vtabel
> Using OpenSSL libs from: openssl/lib/ossl-modules
> Loading base provider...
> Loading fips provider...
> Loading fips provider failed! => FAIL
> - Detailed error: init fail
> call_load_ssl 0

Please provide assistance in resolving this issue. The attached sample code can be used to reproduce the problem.
[test5.tar.gz](https://github.com/user-attachments/files/17006212/test5.tar.gz)

**Comment by paulidale** [comment]
I'd suggest using a second library context for one of the two loads.

**Reply by absrivas-in** [reply]
Sorry, i didn't understand this !

**Reply by absrivas-in** [reply]
This is not only use case where we are seeing this issue,
Imagine we have two shared library for different purpose `crypto operation` and another is `ssl operations`
We are linking `crypto operation` shared library in compile time to our main program.

and for `ssl operations` we are opening shard library 
`void *handle = dlopen("libsecssl.so", RTLD_LAZY | RTLD_DEEPBIND);`
than also we are facing same issue.

**Reply by paulidale** [reply]
The libraries you rely on really should be creating their own library contexts and using them completely independently.
This was one of the fundamental reason for creating library contexts.

If you have two required libraries that both rely on OpenSSL, they must either share the default library context (which is what you are saying is broken) or they need to create their own library contexts and only use them.

My suggestion is to talk to the upstream libraries and ask them nicely to use a library context that they've created.  Raise a bug if required.  There is no way that the OpenSSL project can or will cater for you dependency's needs here.

**Reply by absrivas-in** [reply]
If customer wants they can run `crypto operation` in fips and `ssl operation in fips mode` 
If sharing the libctx is not an option i think !

Are you saying that sdk.c is creating OPENSSL libctx in wrong way ?  
Can you please give some excerpt of the code based on the attached sample code.

Or you are indicating wrong with shared library creation ?

Once you provide the excerpt sample code with that 
`static OSSL_FUNC_get_entropy_fn *c_get_entropy = NULL;`
will above global variable in provider_seeding.c will be NULL in second fips load which will result into resolving this issue ?

**Reply by paulidale** [reply]
Please read the documentation.

This is fairly well covered there.

I am not employed by the project, any assistance I'm giving is exactly that -- a gift.

**Reply by absrivas-in** [reply]
Hi @paulidale ,
Are you referring to following doc ? 
https://docs.openssl.org/3.2/man7/ossl-guide-libraries-introduction/#library-contexts
I could not find anything which can help me here.

We are already creating different library context and doing solitude operations based on the types operation crypto or ssl.

I provided the sample code so that it should be easy to reproduce and debug.
I hope you were able to download and go through the tarball file and Please tell at which file you were referring is wrong.

I am blessed that i have you for the guidance (-- gift) and definitely trust that we will be able to solve this issue together.

**Reply by mattcaswell** [reply]
I don't actually think a second libctx helps in this scenario. The allocation of `c_get_entropy` is failing this check:

```` C
#define set_func(c, f) \
    do { if (c == NULL) c = f; else if (c != f) return 0; } while (0)
````
Where `set_func` is called like this:  `set_func(c_get_entropy, OSSL_FUNC_get_entropy(fns));`

This can only fail if the function pointer supplied from libcrypto for the `get_entropy` callback is *different* between two different invocations. This implies there are two different `libcrypto` instances loaded at the same time - both using the same fips.so. If there was one instance of libcrypto loaded and two different parts of the code trying to load fips.so, then in that case two different libctxs would be the answer. But two different `libcrypto` instances trying to load a single fips.so instance is just not supported and is bound to fail. See this comment in the code:

https://github.com/openssl/openssl/blob/e88dfd5ee50f9d934edd966369339ee5573c67d4/providers/common/provider_seeding.c#L21-L25

You are seeing the failure because the above referenced sanity check is failing. You are in exactly the scenario that this comment is talking about. This is not a supported scenario.

**Reply by absrivas-in** [reply]
Thanks for response @mattcaswell , 
I already had looked into this comment but i am using single libcrypto static library version 3.0.15 from Openssl in both of the invocation. Hence it should not be applicable in my case.

We already have different two different libctxs.
I had attached the sample code tar bar which is mimicking the issue and i had created the debug version so that you can step into.
Can you please have a look  and let me know in which area i should correct the behaviour.

**Reply by mattcaswell** [reply]
> but i am using single libcrypto static library version 3.0.15 from Openssl in both of the invocation

Are you sure? The failing sanity check implies otherwise! From your description it sounds like you have one section of code statically linked to OpenSSL - and another which is dynamically loaded that is also statically linked to OpenSSL. So that's still two instances of OpenSSL (even if its the same version number).

**Reply by t-j-h** [reply]
What flags are used to dlopen if the second instance is doing dynamic loading under program control at runtime?

**Reply by absrivas-in** [reply]
`void *handle = dlopen("libsecssl.so", RTLD_LAZY);`

**Reply by ShawnShealy** [reply]
I also have encountered the same issue as reported here by @absrivas-in.  In my case, the software I ship is a shared library which needs cryptographic functionality and as such I link in object code of the OpenSSL library (compiled to a static library).  No OpenSSL symbols are exported from my shared library.  As a result, the instance of OpenSSL that I use essentially is my own private copy which is isolated from any other software which may be running within the same process.  Note that the way my software has been using OpenSSL is not new as my software has been in use at customer sites without problems for over a decade.  Recently we’ve started testing my software running in FIPS mode.  This all seemed to work just fine with the OpenSSL 3.0.8 object code existing in my shared library loading the FIPS provider and using it for cryptography.  I tested this and thought all was good, since it worked fine.  However, depending upon the application and customer needs, other shared libraries may be loaded into the same process.  What I found is …. when the customer attempted an operation that by its nature needed another shared library (produced by someone else) which also used OpenSSL 3.X, the customer software failed because the attempt to load the FIPS provider failed for whichever instance of OpenSSL attempted to load the FIPS provider second.  I realize that the comment in the code states “We do not support the scenario of an application linked against multiple versions…” is applicable to this situation.  Note that where I encountered this, both instances were the same version (3.0.8) and both libraries statically linked their open copy of OpenSSL.  While this limitation is mentioned in the comment and hence it is obvious that the author of the comment was aware of it, the implications are nasty.  My software passed testing with the FIPS provider just fine and I believe the other party’s software passed testing with the FIPS provider just fine.  It was only when a user tried to do something that used both of these pieces of software (which are dynamically loaded during execution based on inputs provided to the application) that the failure was discovered.  How are developers of software that uses OpenSSL supposed to discover this sort of limitation before their software reaches their customers?  Is this severe limitation documented anywhere (not counting in the source code)?

**Reply by paulidale** [reply]
There is a workaround: rename the fips.so module to something different (e.g. my-fips.so) and load that instead.

The problem isn't related to the developer not wanting to support this, it just doesn't make sense if you think it through.  Two callers are offering the same callback to do stuff.  They obviously cannot be mixed willy nilly.  Likewise, neither set can be used in preference to the other -- the caller's quite possibly have side effects in their callbacks which they rely on happening.  Moreover, at least under Unix like operating systems, it isn't possible to load a shared library twice in a single process and get two distinct "copies", instead a reference count gets incremented and the same instance is used.  This "problem" boils down to a limitation of the host operating system.  Is OpenSSL expected to document all such limitations?  If you think so, please submit a pull request doing this.

**Reply by ShawnShealy** [reply]
Hi @paulidale,

I thank you for your input.

In it, you wrote:

> Two callers are offering the same callback to do stuff.

I agree with this.

You further wrote:

>   They obviously cannot be mixed willy nilly. Likewise, neither set can be used in preference to the other -- the caller's quite possibly have side effects in their callbacks which they rely on happening.

I agree with this.

And also:

> Moreover, at least under Unix like operating systems, it isn't possible to load a shared library twice in a single process and get two distinct "copies", instead a reference count gets incremented and the same instance is used. This "problem" boils down to a limitation of the host operating system.

I do agree with your description of how shared libraries work (I worked on Unix kernel code for over a decade, so this is unsurprising to me).  However, you seem to imply that this scheme doesn't permit the FIPS provider to maintain separate state for each OpenSSL instance.  I disagree with this conclusion.  For example, it seems like the FIPS provider should, upon noticing that the callbacks aren't 100% identical, realize that this is a different OpenSSL library instance and allocate separate context information within the memory of the process.  By having the FIPS provider use separate context information for each OpenSSL library instance existing within the process, the FIPS provider can service these OpenSSL library instances independently.  The FIPS provider's initialization function (OSSL_provider_init()) should return a different value via the provctx parameter; in fact the purpose of this parameter is described as "to allow the provider multiple simultaneous uses", which seems like what we want.  Part of this context information should include the callback addresses so that the provider uses the appropriate callback from the OpenSSL library instance.  This provctx value is passed back in to each of the functions the FIPS provider offers to the OpenSSL library, so these functions are able to discern which OpenSSL library instance is making the call.

Also, regarding:

> There is a workaround: rename the fips.so module to something different (e.g. my-fips.so) and load that instead.

Interestingly, I already am doing essentially this (making a copy of the fips.so file in a different directory) to get past this issue.  But to be honest, I feel this workaround is a bit of a kludge and something I should not have to do.  And to ensure we never have software that fails due to this problem, *every* piece of software that uses OpenSSL would need to implement this kludge (unless they can be sure that there won't be a user of another OpenSSL library instance within the same process).  Getting everyone to implement this workaround would be a bar too high, I suspect.

Regarding:

> Is OpenSSL expected to document all such limitations? If you think so, please submit a pull request doing this.

Honestly, I think it would be preferable to remove the limitation in the software, vs. documenting the limitation.  My software (like other software, I am sure) has used OpenSSL since before the introduction of providers.  Apart from simply upgrading to using a newer version of OpenSSL, my software's usage of OpenSSL hasn't changed.  However, suddenly it starts failing in the presence of customer environments where the system is configured to use the FIPS provider.  I think it is unfortunate that the FIPS provider coder decided to error out in the face of multiple OpenSSL library instances rather than supporting multiple simultaneous uses.

What are your thoughts?

**Reply by paulidale** [reply]
Feel free to submit a pull request making the changes you suggest.  Otherwise, this will have to wait for someone to be allocated time to work on it or someone else to volunteer to do the work.  It ought to be possible to store the callbacks in the provider's context rather than as globals but there will be some functions where the provider context isn't available and adding it as an argument would be very invasive (the `malloc` wrapper e.g.).

**Reply by t8m** [reply]
I assume you would have to use a thread-local storage for the pointer to the current provider context and always set it appropriately when entering the fips provider. Not impossible to do but it would be quite invasive change. Certainly not something for backport to stable branches.

**Reply by absrivas-in** [reply]
Thanks @ShawnShealy 
@paulidale  can we move this discussion to as a bug ?

**Reply by absrivas-in** [reply]
Gentle reminder to consider this as a bug,

**Reply by t8m** [reply]
This is not going to be considered as a bug. As it would require quite deep redesign and refactoring, it is a feature request at the best.

**Reply by t8m** [reply]
I would suggest you to open a new issue with feature request linking this discussion from there.

**Reply by paulidale** [reply]
I've created a feature issue for this: #25553
